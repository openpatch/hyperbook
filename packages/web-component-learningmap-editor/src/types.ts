import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface CompletionNeed {
  id: string;
  source?: string;
  target?: string;
}

export interface Completion {
  needs?: CompletionNeed[];
  optional?: CompletionNeed[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  background?: BackgroundConfig;
  edgeConfig?: EdgeConfig;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
