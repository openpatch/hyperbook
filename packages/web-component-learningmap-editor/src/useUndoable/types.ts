export type ActionType =
  | "undo"
  | "redo"
  | "update"
  | "reset"
  | "resetInitialState";

export type HistoryLimit = number | "infinium" | "infinity";

export type MutationBehavior =
  | "mergePastReversed"
  | "mergePast"
  | "destroyFuture"
  | "keepFuture";

export interface Action<T> {
  type: ActionType;
  payload?: T;
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
  ignoreAction?: boolean;
}

export interface State<T> {
  past: T[];
  present: T;
  future: T[];
}

export interface Options {
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
}

export type UseUndoable<T> = [
  T,
  (
    payload: T | ((oldValue: T) => T),
    behavior?: MutationBehavior,
    ignoreAction?: boolean,
  ) => void,
  {
    past: T[];
    future: T[];

    undo: () => void;
    canUndo: boolean;
    redo: () => void;
    canRedo: boolean;

    reset: (initialState?: T) => void;
    resetInitialState: (newInitialState: T) => void;
    static_setState: (
      payload: T,
      behavior?: MutationBehavior,
      ignoreAction?: boolean,
    ) => void;
  },
];
