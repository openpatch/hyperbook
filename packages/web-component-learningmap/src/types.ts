import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface Completion {
  needs?: string[];
  optional?: string[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface Settings {
  title?: string;
  background?: BackgroundConfig;
  language?: string;
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  settings: Settings;
  version: number;
}

export interface RoadmapState {
  nodes: Record<string, { state: string }>;
  x: number;
  y: number;
  zoom: number;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
