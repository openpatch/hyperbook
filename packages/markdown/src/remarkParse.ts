// @ts-nocheck
import { Root } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { Plugin } from "unified";
import micromarkSubSup from "./micromarkSubSup";
import mdastSubSup from "./mdastSubSup";

const remarkParse: Plugin<[], string, Root> = function () {
  const self = this;

  self.parser = function (doc) {
    const extensions = self.data("micromarkExtensions") || [];
    const mdastExtensions = self.data("fromMarkdownExtensions") || [];

    return fromMarkdown(doc, {
      ...self.data("settings"),
      extensions: [micromarkSubSup, ...extensions],
      mdastExtensions: [mdastSubSup, ...mdastExtensions],
    });
  };
};

export default remarkParse;
