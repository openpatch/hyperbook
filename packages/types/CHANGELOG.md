# @hyperbook/types

## 0.0.1

### Patch Changes

- [#316](https://github.com/openpatch/hyperbook/pull/316) [`cedb551`](https://github.com/openpatch/hyperbook/commit/cedb55191fd025b5a214df406a53cbab5d1b1bc1) Thanks [@mikebarkmin](https://github.com/mikebarkmin)! - Extract Hyperbook types into a separate package
