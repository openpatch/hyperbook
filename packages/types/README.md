# @hyperbook/types

## Installation

```sh
yarn add @hyperbook/types
# or
npm i @hyperbook/types
```
