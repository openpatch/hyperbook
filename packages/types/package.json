{
  "name": "@hyperbook/types",
  "version": "0.5.0",
  "author": "Mike Barkmin",
  "homepage": "https://github.com/openpatch/hyperbook#readme",
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.mjs",
  "types": "dist/index.d.ts",
  "typings": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "require": "./dist/index.cjs.js",
      "default": "./dist/index.esm.mjs"
    }
  },
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/openpatch/hyperbook.git",
    "directory": "packages/types"
  },
  "bugs": {
    "url": "https://github.com/openpatch/hyperbook/issues"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "version": "pnpm build",
    "lint": "tsc --noEmit",
    "build": "pnpm build:pkg && pnpm build:types",
    "build:pkg": "node ../../scripts/build.mjs",
    "build:types": "tsc --project tsconfig.build.json --declaration --emitDeclarationOnly"
  },
  "dependencies": {}
}
