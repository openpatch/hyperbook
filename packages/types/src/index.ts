type ElementConfig = {
  version?: string;
};

export type LinkWithLinks = {
  label: string;
  icon?: string;
  links: Link[];
};

export type LinkWithHref = {
  label: string;
  icon?: string;
  href: string;
};

export type Link = LinkWithHref | LinkWithLinks;

export type Navigation = {
  next: HyperbookPage | null;
  previous: HyperbookPage | null;
  current: HyperbookPage | null;
  pages: HyperbookPage[];
  sections: HyperbookSection[];
};

export type Term = {
  name: string;
  href: string;
};

export type Glossary = Record<string, Term[]>;

export type Language = "de" | "en" | "fr" | "es" | "it" | "pt" | "nl";

export type HyperbookPageFrontmatter = {
  name: string;
  lang?: Language;
  description?: string;
  keywords?: string[];
  index?: number;
  hide?: boolean;
  toc?: boolean;
};

export type HyperbookSectionFrontmatter = HyperbookPageFrontmatter & {
  virtual?: boolean;
  expanded?: boolean;
};

export type HyperbookPage = HyperbookPageFrontmatter & {
  isEmpty?: boolean;
  href?: string;
  repo?: string;
};

export type HyperbookSection = HyperbookSectionFrontmatter & {
  isEmpty?: boolean;
  href?: string;
  pages: HyperbookPage[];
  sections: HyperbookSection[];
  repo?: string;
};

export type HyperbookFrontmatter =
  | HyperbookPageFrontmatter
  | HyperbookSectionFrontmatter;

export type HyperbookJson = {
  name: string;
  description?: string;
  logo?: string;
  author?: {
    name?: string;
    url?: string;
  };
  font?: string;
  fonts?: {
    heading?: string;
    body?: string;
    code?: string;
  };
  colors?: {
    /**
     * @format color-hex
     */
    brand?: string;
    /**
     * @format color-hex
     */
    brandDark?: string;
    /**
     * @format color-hex
     */
    brandText?: string;
    inverted?: boolean;
  };
  basePath?: string;
  license?: string;
  language?: Language;
  repo?: string;
  links?: Link[];
  template?: string;
  elements?: {
    bookmarks?: false | ElementConfig;
    excalidraw?: ElementConfig & {
      autoZoom?: boolean;
      center?: boolean;
      aspectRation?: string;
      edit?: boolean;
    };
  };
};

export type LanguageString = Partial<Record<Language, string>>;

export type HyperlibraryJson = {
  name: string | LanguageString;
  library: {
    src: string;
    basePath: string;
    icon?: string;
    name?: string | LanguageString;
  }[];
  basePath?: string;
};

export type Hyperproject = {
  name: string | LanguageString;
  basePath?: string;
  icon?: string;
  src: string;
} & ({ type: "book" } | { type: "library"; projects: Hyperproject[] });
