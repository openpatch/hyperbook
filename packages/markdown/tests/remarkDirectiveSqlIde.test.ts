import { HyperbookContext } from "@hyperbook/types";
import { describe, expect, it } from "vitest";
import rehypeStringify from "rehype-stringify";
import remarkParse from "remark-parse";
import remarkToRehype from "remark-rehype";
import rehypeFormat from "rehype-format";
import { unified, PluggableList } from "unified";
import remarkDirective from "remark-directive";
import remarkDirectiveRehype from "remark-directive-rehype";
import remarkDirectiveSqlIde from "../src/remarkDirectiveSqlIde";
import { ctx } from "./mock";

export const toHtml = (md: string, ctx: HyperbookContext) => {
  const remarkPlugins: PluggableList = [
    remarkDirective,
    remarkDirectiveRehype,
    remarkDirectiveSqlIde(ctx),
  ];

  return unified()
    .use(remarkParse)
    .use(remarkPlugins)
    .use(remarkToRehype)
    .use(rehypeFormat)
    .use(rehypeStringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
    })
    .processSync(md);
};

describe("remarkDirectiveSqlIde", () => {
  it("should transform", async () => {
    expect(
      toHtml(
        `:::sqlide{height=500}

\`\`\`mysql Statements.sql

SELECT * from fluss;

\`\`\`

\`\`\`md A Hint

Do this

\`\`\`
:::

`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
});
