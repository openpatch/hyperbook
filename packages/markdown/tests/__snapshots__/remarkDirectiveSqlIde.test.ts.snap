// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkDirectiveSqlIde > should transform 1`] = `
"
<div class="directive-sqlide">
  <iframe class="player" srcdoc="<script>window.jo_doc = window.frameElement.textContent;</script><script src='https://embed.learn-sql.de/include/js/includeide/includeIDE.js'></script>" frameborder="0" height="500">{'id': 'fa3f84c5', 'databaseURL': 'https://embed.learn-sql.de/include/assets/databases/world1.sqLite'}
    <script type="plain/text" title="Statements.sql">
SELECT * from fluss;
</script>
    <script type="plain/text" title="A Hint" data-type="hint">
Do this
</script>
    <style>
.joe_javaOnlineDiv {
  box-shadow: none;
  margin: 0!important;
  top: 0!important;
  width: 100%!important;
  height: calc(100% - 5px) !important;
  border: none !important;
}
</style>
  </iframe>
  <div class="menu"><button onclick="hyperbook.sqlide.openFullscreen(this)">Fullscreen</button></div>
</div>
"
`;
